{
 "metadata": {
  "name": "",
  "signature": "sha256:fc9ddde3e5781d22fd86b9dd69a5f4252dc863c96d7c2a95d415724aea57c879"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import sys,os,csv,time,math,random,copy\n",
      "import numpy as np\n",
      "import Bio.PDB\n",
      "import pymysql as db\n",
      "import pymysql.cursors\n",
      "from warnings import filterwarnings,resetwarnings\n",
      "filterwarnings('ignore', category = db.Warning)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "query = \"\"\"\n",
      "SELECT\n",
      "/*SLABEL*/a.label as slabel,\n",
      "/*ID*/a.structid,\n",
      "/*METHOD*/IF(c.method IS NULL,d.method,c.method) as method,\n",
      "/*STRUCTURE*/c.resolution,\n",
      "/*MODEL*/d.mpqs,\n",
      "/*CHAIN*/b.biounit,b.model,b.chain,b.unp,b.hybrid,\n",
      "/*RESIDUE*/a.seqid as seqid,a.icode,a.rescode,a.ss,a.phi,a.psi,a.tco,a.kappa,a.alpha,a.x,a.y,a.z,\n",
      "/*PFAM*/p.acc as pfamid,p.name as pfam_domain,p.description as pfam_desc,p.evalue as pfam_evalue,\n",
      "/*ALIGN*/h.chain_seqid as aln_chain_seqid,h.trans_seqid as aln_trans_seqid,j.perc_identity,\n",
      "/*TRANS*/i.transcript as enst,i.gene as ensg,i.protein as ens_prot,i.seqid as ens_prot_seqid,i.rescode as ens_prot_aa,\n",
      "/*DLABEL*/g.label as dlabel,\n",
      "/*VARIANT*/(e.slabel='uniprot-pdb' and e.dlabel='1kg' and f.label='1kg' AND f.consequence LIKE '%%missense_variant%%') as issnp,\n",
      "/*VARIANT*/g.name as snpid,g.chr,g.start,g.end,g.hgnc_gene,g.ens_gene,g.aa as anc_allele,g.ref_allele,g.alt_allele,g.maf,g.amr_af,g.asn_af,g.eur_af,g.afr_af,\n",
      "/*CONSEQUENCE*/f.gc_id,f.transcript as vep_trans,f.protein as vep_prot,f.protein_pos as vep_prot_pos,f.ref_codon,f.alt_codon,f.ref_amino_acid as vep_ref_aa,f.alt_amino_acid as vep_alt_aa,\n",
      "/*CONSEQUENCE*/f.consequence,f.polyphen,f.sift,f.biotype\n",
      "FROM Residue as a\n",
      "LEFT JOIN Chain as b \n",
      "ON a.label=b.label AND a.structid=b.structid AND a.biounit=b.biounit AND a.model=b.model AND a.chain=b.chain\n",
      "LEFT JOIN Structure as c\n",
      "ON b.label=c.label AND b.structid=c.pdbid\n",
      "LEFT JOIN Model as d\n",
      "ON b.label=d.label AND b.structid=d.modelid\n",
      "LEFT JOIN GenomicIntersection as e\n",
      "ON a.label=e.slabel AND a.structid=e.structid AND a.chain=e.chain AND a.seqid=e.seqid\n",
      "LEFT JOIN GenomicConsequence as f\n",
      "ON e.dlabel=f.label AND e.gc_id=f.gc_id\n",
      "LEFT JOIN GenomicData as g\n",
      "ON f.label=g.label AND f.chr=g.chr AND f.start=g.start AND f.end=g.end AND f.name=g.name\n",
      "LEFT JOIN Alignment as h USE INDEX(PRIMARY)\n",
      "ON a.label=h.label AND a.structid=h.structid AND a.chain=h.chain AND a.seqid=h.chain_seqid #AND f.transcript=h.transcript\n",
      "LEFT JOIN Transcript as i USE INDEX(PRIMARY)\n",
      "ON h.label=i.label AND h.transcript=i.transcript AND h.trans_seqid=i.seqid\n",
      "LEFT JOIN AlignmentScore as j\n",
      "ON h.label=j.label AND h.structid=j.structid AND h.chain=j.chain AND h.transcript=j.transcript\n",
      "LEFT JOIN pfam as p\n",
      "ON a.structid=p.pdbid AND a.chain=p.chain AND a.seqid BETWEEN p.seqstart AND p.seqend\n",
      "WHERE a.label='uniprot-pdb'\n",
      "AND b.label='uniprot-pdb'\n",
      "AND (g.label IS NULL OR g.label='1kg')\n",
      "AND (f.transcript IS NULL OR f.transcript=h.transcript)\n",
      "AND (c.method LIKE '%%nmr%%' OR b.biounit>0 OR NOT ISNULL(d.modelid))\n",
      "AND a.structid=%s AND a.biounit=%s\n",
      "ORDER BY b.unp,a.seqid DESC;\n",
      "\"\"\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def connect(retry=True):\n",
      "  try:\n",
      "    if connect.con: return connect.con.cursor()\n",
      "    connect.con = db.connect(host='gwar-dev',user='mike',passwd='cheezburger',\n",
      "                             db='pdbmap_v10',cursorclass=db.cursors.SSDictCursor)\n",
      "    return connect.con.cursor()\n",
      "  except db.Error as e:\n",
      "    msg = \"There was an error connecting to the database: %s\\n\"%e\n",
      "    sys.stderr.write(msg)\n",
      "    if retry:\n",
      "      msg = \"Waiting 30-300s and retrying...\\n\"\n",
      "      sys.stderr.write(msg)\n",
      "      time.sleep(random.randint(30,300)) # Wait 30-300 seconds and retry\n",
      "      return connect(retry=False)\n",
      "    else:\n",
      "      msg = \"Database connection unsuccessful: %s\\n\"%e\n",
      "      sys.stderr.write(msg)\n",
      "      raise\n",
      "connect.con = None"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def secure_query(query,qvars=None,cursorclass='SSDictCursor'):\n",
      "    \"\"\" Executes queries using safe practices \"\"\"\n",
      "    c = connect()\n",
      "    filterwarnings('ignore', category = db.Warning)\n",
      "    try:\n",
      "      if qvars: c.execute(query,qvars)\n",
      "      else:     c.execute(query)\n",
      "    except Exception as e:\n",
      "      msg  = \"ERROR (PDBMapIO) Secure query failed; Exception: %s; \"%str(e)\n",
      "      msg += \" Provided query: %s\"%query\n",
      "      msg += \" Provided args: %s\"%str(qvars)\n",
      "      if \"_last_executed\" in dir(c):\n",
      "        msg += \" Executed Query: %s\"%c._last_executed\n",
      "      raise Exception(msg)\n",
      "    resetwarnings()\n",
      "    for row in c:\n",
      "      yield row\n",
      "    c.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def load_structure(structid,biounit=0):\n",
      "  global query\n",
      "  q = secure_query(query,qvars=(structid,biounit),cursorclass='DictCursor')\n",
      "  # Construct the Bio.PDB.Structure\n",
      "  s = None\n",
      "  for row in q:\n",
      "    if not s:\n",
      "        s = Bio.PDB.Structure.Structure(row['structid'])\n",
      "        s.biounit    = row['biounit']\n",
      "        s.resolution = row['resolution'] if row['resolution'] else None\n",
      "        s.mpqs       = row['mpqs'] if row['mpqs'] else None\n",
      "        s.method     = row['method']\n",
      "        s.snpmapper  = {}\n",
      "    m = row['model']\n",
      "    if m not in s: s.add(Bio.PDB.Model.Model(m))\n",
      "    m = s[m]\n",
      "    c = row['chain']\n",
      "    if c not in m: m.add(Bio.PDB.Chain.Chain(c))\n",
      "    c = m[c]\n",
      "    c.unp         = row['unp']\n",
      "    c.transcript  = row['enst']\n",
      "    c.gene        = row['ensg']\n",
      "    c.perc_identity = row['perc_identity']\n",
      "    c.hybrid  = row['hybrid']\n",
      "    r = (' ',row['seqid'], ' ')\n",
      "    c.add(Bio.PDB.Residue.Residue(r,row['rescode'],' '))\n",
      "    r = c[r]\n",
      "    r.seqid = row['seqid']\n",
      "    r.coords = (row['x'],row['y'],row['z'])\n",
      "    r.pfam   = (row['pfamid'],row['pfam_domain'],row['pfam_desc'],row['pfam_evalue'])\n",
      "    r.ss     = row['ss']\n",
      "    r.angles = (row['phi'],row['psi'],row['tco'],row['kappa'],row['alpha'])\n",
      "    if c.unp not in s.snpmapper:\n",
      "        s.snpmapper[c.unp] = np.array([[None for j in range(19)] for i in range(r.seqid)])\n",
      "    if not np.any(s.snpmapper[c.unp][r.seqid-1]):\n",
      "        snp  = [row['issnp'],row['snpid'],row['chr'],row['start'],row['end'],row['hgnc_gene'],row['ens_gene']]\n",
      "        snp += [row['anc_allele'],row['ref_allele'],row['alt_allele']]\n",
      "        snp += [row['maf'],row['amr_af'],row['asn_af'],row['eur_af'],row['afr_af']]\n",
      "        snp += [row['ref_codon'],row['alt_codon'],row['vep_ref_aa'],row['vep_alt_aa']]\n",
      "        s.snpmapper[c.unp][r.seqid-1] = snp\n",
      "    r.snp = s.snpmapper[c.unp][r.seqid-1]\n",
      "  return s"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def permute(ts,n=10):\n",
      "    ts = copy.deepcopy(s.snpmapper)\n",
      "    for i in range(n):\n",
      "        for unp in s.snpmapper.keys():\n",
      "            np.random.shuffle(s.snpmapper[unp])\n",
      "        yield s\n",
      "    s.snpmapper = ts"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_snps(s):\n",
      "    return [(r.seqid,r.snp[1]) for r in s[1]['A'] if r.snp[0]]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t0 = time.time()\n",
      "s = load_structure('1pl8',biounit=1)\n",
      "res = [get_snps(x) for x in permute(s,1000)]\n",
      "print \"1000 permutations required %2.2fs\"%(time.time()-t0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for i,r1 in enumerate(res[5:]):\n",
      "    for j,r2 in enumerate(res[5:]):\n",
      "        if i != j:\n",
      "            print np.all(i==j)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}