{
 "metadata": {
  "name": "",
  "signature": "sha256:b79b4dce1ca2340aa5b7b43ad890fabe6432efe7cd51313ecf59a4c5b8e4d20f"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import sys,os,csv,time,math,random,copy\n",
      "import numpy as np\n",
      "import Bio.PDB\n",
      "import pymysql as db\n",
      "import pymysql.cursors\n",
      "from warnings import filterwarnings,resetwarnings\n",
      "filterwarnings('ignore', category = db.Warning)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 91
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "THIS QUERY WILL NOT WORK\n",
      "I need to write a new query which is a composite of the one below (which contains nearly all fields)\n",
      "and the one in sliding_sphere.py which uses left joins to query all residues, but attaches SNP\n",
      "data to those rows that are mapped to SNPs.\n",
      "\n",
      "query = \"\"\"\n",
      "SELECT\n",
      "/*LABELS*/a.label as dlabel,d.label as slabel,\n",
      "/*VARIANT*/a.name,a.chr,a.start,a.end,a.hgnc_gene,b.gc_id,\n",
      "/*CONSEQUENCE*/b.transcript as vep_trans,b.protein as vep_prot,b.protein_pos as vep_prot_pos,\n",
      "/*CONSEQUENCE*/b.ref_amino_acid as vep_ref_aa,b.alt_amino_acid as vep_alt_aa,b.consequence,\n",
      "/*ALIGN*/h.chain_seqid as aln_chain_seqid,h.trans_seqid as aln_trans_seqid,j.perc_identity,\n",
      "/*TRANS*/i.gene as ens_gene,i.transcript as ens_trans,i.protein as ens_prot,i.seqid as ens_prot_seqid,i.rescode as ens_prot_aa,\n",
      "/*RESIDUE*/d.structid,d.seqid as rid,d.rescode,d.x,d.y,d.z,\n",
      "/*CHAIN*/e.biounit,e.model,e.chain,e.unp,\n",
      "/*STRUCTURE*/f.pdbid,f.method as smethod,f.resolution,\n",
      "/*MODEL*/g.modelid,g.method as mmethod,g.mpqs\n",
      "FROM GenomicData AS a\n",
      "INNER JOIN GenomicConsequence AS b\n",
      "ON a.label=b.label AND a.chr=b.chr AND a.start=b.start AND a.end=b.end AND a.name=b.name\n",
      "INNER JOIN GenomicIntersection AS c\n",
      "ON b.label=c.dlabel AND b.gc_id=c.gc_id\n",
      "INNER JOIN Residue AS d\n",
      "ON c.slabel=d.label AND c.structid=d.structid AND c.chain=d.chain AND c.seqid=d.seqid\n",
      "INNER JOIN Chain as e\n",
      "ON d.label=e.label AND d.structid=e.structid AND d.biounit=e.biounit AND d.model=e.model AND d.chain=e.chain\n",
      "LEFT JOIN Structure as f\n",
      "ON e.label=f.label AND e.structid=f.pdbid\n",
      "LEFT JOIN Model as g\n",
      "ON e.label=g.label AND e.structid=g.modelid\n",
      "INNER JOIN Alignment as h USE INDEX(PRIMARY)\n",
      "ON d.label=h.label AND d.structid=h.structid AND d.chain=h.chain AND d.seqid=h.chain_seqid AND h.transcript=b.transcript\n",
      "INNER JOIN Transcript as i USE INDEX(PRIMARY)\n",
      "ON h.label=i.label AND h.transcript=i.transcript AND h.trans_seqid=i.seqid\n",
      "INNER JOIN AlignmentScore as j\n",
      "ON h.label=j.label AND h.structid=j.structid AND h.chain=j.chain AND h.transcript=j.transcript\n",
      "WHERE a.label='1kg'\n",
      "AND b.label='1kg'\n",
      "AND c.dlabel='1kg' AND c.slabel='uniprot-pdb'\n",
      "AND d.label='uniprot-pdb'\n",
      "AND e.label='uniprot-pdb'\n",
      "AND h.label='uniprot-pdb'\n",
      "AND i.label='uniprot-pdb'\n",
      "AND j.label='uniprot-pdb'\n",
      "AND d.structid=%s AND d.biounit=%s\n",
      "# Limit to biological assemblies for experimental structures\n",
      "AND (f.method LIKE '%%nmr%%' OR e.biounit>0 OR NOT ISNULL(g.modelid))\n",
      "# Limit to missense SNPs\n",
      "AND b.consequence LIKE '%%missense_variant%%'\n",
      "ORDER BY a.name,a.chr,a.start,a.end,d.structid,d.biounit,d.chain,d.seqid;\n",
      "# Do not specify structure/model labels. One will be NULL in each row.\n",
      "# Structure/model labels will still match the chain labels, which are specified.\n",
      "\"\"\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 141
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def connect(retry=True):\n",
      "  try:\n",
      "    con = db.connect(host='gwar-dev',user='mike',passwd='cheezburger',\n",
      "                          db='pdbmap_v10',cursorclass=db.cursors.SSDictCursor)\n",
      "    return con.cursor()\n",
      "  except db.Error as e:\n",
      "    msg = \"There was an error connecting to the database: %s\\n\"%e\n",
      "    sys.stderr.write(msg)\n",
      "    if retry:\n",
      "      msg = \"Waiting 30-300s and retrying...\\n\"\n",
      "      sys.stderr.write(msg)\n",
      "      time.sleep(random.randint(30,300)) # Wait 30-300 seconds and retry\n",
      "      return _connect(retry=False)\n",
      "    else:\n",
      "      msg = \"Database connection unsuccessful: %s\\n\"%e\n",
      "      sys.stderr.write(msg)\n",
      "      raise"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 142
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def secure_query(query,qvars=None,cursorclass='SSDictCursor'):\n",
      "    \"\"\" Executes queries using safe practices \"\"\"\n",
      "    c = connect()\n",
      "    filterwarnings('ignore', category = db.Warning)\n",
      "    try:\n",
      "      if qvars: c.execute(query,qvars)\n",
      "      else:     c.execute(query)\n",
      "    except Exception as e:\n",
      "      msg  = \"ERROR (PDBMapIO) Secure query failed; Exception: %s; \"%str(e)\n",
      "      msg += \" Provided query: %s\"%query\n",
      "      msg += \" Provided args: %s\"%str(qvars)\n",
      "      if \"_last_executed\" in dir(c):\n",
      "        msg += \" Executed Query: %s\"%c._last_executed\n",
      "      raise Exception(msg)\n",
      "    resetwarnings()\n",
      "    for row in c:\n",
      "      yield row"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 143
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def load_structure(structid,biounit=0):\n",
      "  global query\n",
      "  q = secure_query(query,qvars=(structid,biounit),cursorclass='DictCursor')\n",
      "  # Construct the Bio.PDB.Structure\n",
      "  s = None\n",
      "  for row in q:\n",
      "    if not s:\n",
      "        s = Bio.PDB.Structure.Structure(row['structid'])\n",
      "        s.biounit    = row['biounit']\n",
      "        s.resolution = row['resolution'] if row['resolution'] else None\n",
      "        s.mpqs       = row['mpqs'] if row['mpqs'] else None\n",
      "        s.method     = row['smethod'] if row['smethod'] else row['mmethod']\n",
      "    m = row['model']\n",
      "    if m not in s: s.add(Bio.PDB.Model.Model(m))\n",
      "    m = s[m]\n",
      "    c = row['chain']\n",
      "    if c not in m: m.add(Bio.PDB.Chain.Chain(c))\n",
      "    c = m[c]\n",
      "    c.unp = row['unp']\n",
      "    r = (' ',row['rid'], ' ')\n",
      "    c.add(Bio.PDB.Residue.Residue(r,row['rescode'],' '))\n",
      "    r = c[r]\n",
      "    r.coords = (row['x'],row['y'],row['z'])\n",
      "    print row['structid'],row['biounit'],row['model'],row['chain'],row['unp'],row['rid']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 144
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "load_structure('1pl8',biounit=1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1PL8 1 2 A Q00796 102\n",
        "1PL8 1 1 A Q00796 102\n",
        "1PL8 1 2 B Q00796 102\n",
        "1PL8 1 1 B Q00796 102\n",
        "1PL8 1 2 A Q00796 254\n",
        "1PL8 1 1 A Q00796 254\n",
        "1PL8 1 1 B Q00796 254\n",
        "1PL8 1 2 B Q00796 254\n",
        "1PL8 1 1 A Q00796 152\n",
        "1PL8 1 2 A Q00796 152\n",
        "1PL8 1 1 B Q00796 152\n",
        "1PL8 1 2 B Q00796 152\n",
        "1PL8 1 1 A Q00796 90\n",
        "1PL8 1 2 A Q00796 90\n",
        "1PL8 1 1 B Q00796 90\n",
        "1PL8 1 2 B Q00796 90\n",
        "1PL8 1 1 A Q00796 347\n",
        "1PL8 1 2 A Q00796 347\n",
        "1PL8 1 1 B Q00796 347\n",
        "1PL8 1 2 B Q00796 347\n",
        "1PL8 1 1 A Q00796 321\n",
        "1PL8 1 2 A Q00796 321\n",
        "1PL8 1 1 B Q00796 321\n",
        "1PL8 1 2 B Q00796 321\n",
        "1PL8 1 2 A Q00796 107\n",
        "1PL8 1 1 A Q00796 107\n",
        "1PL8 1 2 B Q00796 107\n",
        "1PL8 1 1 B Q00796 107\n",
        "1PL8 1 1 A Q00796 238\n",
        "1PL8 1 2 A Q00796 238\n",
        "1PL8 1 1 B Q00796 238\n",
        "1PL8 1 2 B Q00796 238\n",
        "1PL8 1 1 A Q00796 356\n",
        "1PL8 1 2 A Q00796 356\n",
        "1PL8 1 1 B Q00796 356\n",
        "1PL8 1 2 B Q00796 356\n",
        "1PL8 1 2 A Q00796 268\n",
        "1PL8 1 1 A Q00796 268\n",
        "1PL8 1 1 B Q00796 268\n",
        "1PL8 1 2 B Q00796 268\n"
       ]
      }
     ],
     "prompt_number": 145
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 104
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}